File: Prob1032/Solution.cpp
// Approach 1:-

class TrieNode {
    TrieNode* children[26];
    bool word;
    public:    
    
    TrieNode() {
        word=false;
        for(int i=0;i<26;i++)
            children[i]=nullptr;
    }
    
    void insert(string& s) {
        TrieNode* temp=this;
        
        for(int i=s.size()-1;i>=0;i--) {
            if(temp->children[s[i]-'a']==nullptr)
                temp->children[s[i]-'a']=new TrieNode();
            
            temp=temp->children[s[i]-'a'];
        }
        
        temp->word=true;
        
        return ;
    }
    
    bool search(vector<char>& prev) {
        TrieNode* temp=this;
        
        for(int i=prev.size()-1;i>=0 && temp;i--) {
            int n=prev[i]-'a';
            
            if(temp->children[n]==nullptr)
                return false;
            
            temp=temp->children[n];
            
            if(temp->word)
                return true;
        }
        
        return false;
    }
};


class StreamChecker {
public:
    vector<char> queryChar;
    TrieNode root;
    StreamChecker(vector<string>& words) {
        for(int i=0;i<words.size();i++)
            root.insert(words[i]);
    }
    
    bool query(char letter) {
        queryChar.push_back(letter);
        return root.search(queryChar);
    }
};

File: Prob1032/Solution.java
// Approach 1:-

class StreamChecker {
    
    class TrieNode {
        boolean isWord=false;
        TrieNode[] next=new TrieNode[26];
    }
    
    TrieNode root=new TrieNode();
    StringBuilder sb = new StringBuilder();

    public StreamChecker(String[] words) {
        createTrie(words);
    }
    
    public boolean query(char letter) {
        sb.append(letter);
        
        TrieNode node=root;
        
        for(int i=sb.length()-1;i>=0 && node!=null;i--) {
            char c=sb.charAt(i);
            node=node.next[c-'a'];
            if(node!=null && node.isWord)
                return true;
        }
        
        return false;
    }
    
    private void createTrie(String[] words) {
        for(String s: words) {
            TrieNode node=root;
            int len=s.length();
            
            for(int i=len-1;i>=0;i--) {
                char c=s.charAt(i);
                if(node.next[c-'a']==null)
                    node.next[c-'a']=new TrieNode();
                
                node=node.next[c-'a'];
            }
            
            node.isWord=true;
        }
    }
}